flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+result to be returned. Given the argument and result types are scalar characters, the related
+`fir::Extended` must be `fir::CharBoxValue`, which is simply a container over two SSA values: one
+for the address, one for the length.
+
+The choice here was to implement Trim with the help of runtime. So what the code is preparing the arguments
+according to the runtime interface. TRIM runtime is described in the next section.
+It takes a `Descriptor&` for the result, and a `const Descriptor&` for the argument. The result can be seen
+as a temporary allocatable that the runtime will allocate.
+The code here is creating a `fir::MutableBoxValue`, which is a class used to deal with allocatables in lowering.
+It also creates a descriptor for the argument with `builder.createBox()`. Note that the two descriptors are created
+differently because we expect the first one to be modified, but not the second one.
+
+The actual runtime call is generated by `Fortran::lower::genTrim` defined in `lib/Lower/CharacterRuntime.cpp`
+After the runtime call, the result descriptor is read to build the resulting `fir::CharBoxValue`
+with the result address and length. Since the result address was allocated, it is added to a clean-up list so that
+it can be deallocated after the statement.
+
+The implementation of the runtime call in lowering looks like:
+
+```
+void Fortran::lower::genTrim(Fortran::lower::FirOpBuilder &builder,
+                             mlir::Location loc, mlir::Value resultBox,
+                             mlir::Value stringBox) {
+  auto trimFunc = getRuntimeFunc<mkRTKey(Trim)>(loc, builder);
+  auto fTy = trimFunc.getType();
+  auto sourceFile = Fortran::lower::locationToFilename(builder, loc);
+  auto sourceLine =
+      Fortran::lower::locationToLineNo(builder, loc, fTy.getInput(3));
+
+  llvm::SmallVector<mlir::Value, 4> args;
+  args.emplace_back(builder.createConvert(loc, fTy.getInput(0), resultBox));
+  args.emplace_back(builder.createConvert(loc, fTy.getInput(1), stringBox));
+  args.emplace_back(builder.createConvert(loc, fTy.getInput(2), sourceFile));
+  args.emplace_back(builder.createConvert(loc, fTy.getInput(3), sourceLine));
+  builder.create<fir::CallOp>(loc, trimFunc, args);
+}
+```
+
+The key point is `getRuntimeFunc<mkRTKey(Trim)>(loc, builder)` that builds the FIR signature for the runtime
+function automatically. The name passed to `mkRTKey` must be the same as the one inside `RTNAME` when declaring
+the function in the runtime headers. The runtime header must be included in the current file to use `getRuntimeFunc<>`
+(note the `#include "../../runtime/character.h"` at the top of the file). So at least the runtime API must be designed before adding
+the support in lowering.
+
+Then, the source file name and line number are lowered from the current location so that they can be passed to the runtime.
+In general, runtime that may fail (like here if the allocation was to fail) takes source file information.
```

Comment:
I would say "runtime calls that may fail (like here if the allocation were to fail) take source"